Installation fixes

compor opened the pull request:
This PR is based on issues that came upon a "clean" Ubuntu 21.10 Docker image:

- Update the prerequisites list and the README file for this.
- Fix use of base path (`--base-path` commandline option) used for:
  - Utility scripts installation
  - Makefile template generation  
 
  This allows installation of the Popcorn compiler without having to use the repo's top dir as the CWD.  
  Ideally, the `--base-path` value should be autodetected, but I had no time to make this happen.
 - Add hunk to LLVM 9 patch that triggers a failed compilation due to a missing standard header inclusion (`#include <limits>`).
   This is part of the benchmarking code of the project, thus inconsequential for the core Popcorn compiler.

compor committed:
Update prerequisites

compor committed:
Fix base path use for Makefile template generation

compor committed:
Fix base path use for scripts dir installation

compor committed:
Add hunk to LLVM patch

compor committed:
Update README for prerequisites

compor commented:
@xjtuwxg this is not urgent, but a look into it would be appreciated :)

xjtuwxg commented:
Thanks for the pull request. Seems like you want to upgrade the building script to a later Ubuntu version. Do you also have an update on the [Dockerfile](https://github.com/ssrg-vt/popcorn-compiler/blob/main/docker/Dockerfile)? Here is a [link](https://popcornlinux-doc.readthedocs.io/en/latest/build_compiler.html#build-popcorn-compiler-using-docker-recommended) how to use it.  Also, should we choose an LTS Ubuntu version like 20.04 or 22.04 as the base docker image?

I'm having a quick test using your script on a Ubuntu 20.04 base docker image, there seem a lot of warnings during the building process of LLVM/Clang9 (where 18.04 has a perfect building process w/o warnings). But I understand we cannot use the 18.04 docker image forever.

compor committed:
Use correct gcc spec file when cross-compiling tools

compor committed:
Add docker file updates

compor committed:
Appease GCC compilation warnings.

- Provide C++ standard explicitly
- Disable warnings spam due to newer GCC version

compor commented:
Hi again @xjtuwxg,

As per above, here's a more refined breakdown:

- Update Dockerfile to current Ubuntu LTS version 22.04 (Jammy)
- Breakdown for warnings emitted by GCC 11.2 and 11.3:
  - range-loop-construct: This warns that a ranged-for loop copy constructs each element per iteration, e.g.
    `for(auto e: mylist)` instead of `for(const auto &e:  mylist)`.
    These are issues with the LLVM codebase which have been fixed in newer releases like [this][1].
    Hence, this warning is left to be emitted.
  - init-list-lifetime: This is a major contributor to the warning spam (over 1000 warnings).
    However, this is entirely related to how the ArrayRef API is used in LLVM.
     In a lengthy [discussion][2] regarding this issue on Phabricator, it was decided to disable the warning just for that use.
     Thus, a newer LLVM release should avoid emitting this warning.
     For now, we disable the warning.
  - deprecated-copy: This means that according to recent C++ standard conventions when if either copy ctro, operator= or a destructor is explicitly provided, then all 3 must be explicitly provided.  
    This is fixed in newer LLVM releases by adding the missing methods as in [this][3] file.  
    This warning is currently silenced as it emits a lot of warning spam.
  - deprecated-declarations: This is triggered when a deprecated API is used.
    The main culprit here is the use of mallinfo instead of the newer mallinfo2, as also documented [elsewhere][4].
    This warning has been left to be emitted.

Let me know what you think.
Maybe the extra CMake CXX flags in e8bbb2bebcb5fd2f2a85505f7489013c06133489 for the aforementioned warnings could be guarded by a version check, e.g., to revisit them if the LLVM is greater than 9.1.
Not sure if that is something desirable in this project.


[1]: https://github.com/llvm/llvm-project/blob/c441bfb7147d4648ff2c5ccd8a36f5acc1e271fd/llvm/lib/IR/Attributes.cpp#L1185
[2]: https://reviews.llvm.org/D70122?id=229761
[3]: https://github.com/llvm/llvm-project/blob/d9e51e75521d5e33f24a6c1afacae5dbc115f96d/clang/include/clang/Sema/ObjCMethodList.h#L35
[4]: https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/3763
